### PR TITLE
Add references to the particle API, with dereferencing.

### PR DESCRIPTION
**Pull request by shans**
Note that I added references to isSomeSortOfCollection, etc, because it seemed to fit well there. Might be worth discussing.

### CODE REVIEW

File: runtime/particle-execution-context.js [hunk 1]
```diff
@@ -186,6 +193,18 @@ export class ParticleExecutionContext {
...
 
+  getStorageProxy(storageKey, type) {
+    if (!this._keyedProxies[storageKey]) {      
+      this._keyedProxies[storageKey] = new Promise((resolve, reject) => {
+        this._apiPort.GetBackingStore({storageKey, type, callback: proxy => {
+          this._keyedProxies[storageKey] = proxy;
+          resolve(proxy);
+        }});
+      });
+      return this._keyedProxies[storageKey];
```
[the remark this answers is not in the document]
Comment (shans) [reply runtime/particle-execution-context.js]: These are the only proxies that know their storage keys. The other ones are only aware of the endpoints of their connected store.

File: runtime/ts/type.ts [hunk 2]
```diff
@@ -206,12 +206,15 @@ export class Type {
...
+    if (this.isReference) {
+      return this.referenceReferredType;
+    }
     return null;
   }
 
   // TODO: naming is hard
   isSomeSortOfCollection() {
-    return this.isCollection || this.isBigCollection;
+    return this.isCollection || this.isBigCollection || this.isReference;
```
[the remark this answers is not in the document]
Comment (shans) [reply runtime/ts/type.ts]: Not sure what that means, might be worth chatting through F2F?

File: runtime/particle-execution-context.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mykmartin) [reply runtime/particle-execution-context.js]: Ah ok, SGTM.

File: runtime/ts/type.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mykmartin) [reply runtime/ts/type.ts]: The referred type could be a Collection, a BigCollection or a Variable, so shouldn't this be something like:
```
if (this.isCollection || this.isBigCollection) return true;
if (this.isReference) return this.referredType.isSomeSortOfCollection();
```
?

[the remark this answers is not in the document]
Comment (mykmartin) [reply runtime/ts/type.ts]: As discussed, we can change this to isTypeContainer (or maybe hasElement/UnderylingType).

File: runtime/handle.js [hunk 3]
```diff
@@ -142,6 +142,19 @@ class Collection extends Handle {
     }
   }
 
+    /** @method async get(id)
+   * Returns the Entitiy specified by id contained by the handle, or null if this id is not
+   * contained by the handle.
+   * throws: Error if this handle is not configured as a readable handle (i.e. 'in' or 'inout')
+     in the particle's manifest.
+   */
```
[the remark this answers is not in the document]
Comment (shans) [reply runtime/handle.js]: fixed

File: runtime/handle.js [hunk 4]
```diff
@@ -142,6 +142,19 @@ class Collection extends Handle {
     }
   }
 
+    /** @method async get(id)
+   * Returns the Entitiy specified by id contained by the handle, or null if this id is not
```
[the remark this answers is not in the document]
Comment (shans) [reply runtime/handle.js]: Fixed.

File: runtime/ts/reference.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {ParticleExecutionContext} from '../particle-execution-context.js';
+import {Type} from './type.js';
+import {handleFor} from '../handle.js';
+
+export class Reference {
+  public entity = null;
```
[the remark this answers is not in the document]
Comment (shans) [reply runtime/ts/reference.ts]: Apparently it doesn't? I'm not sure what type exactly I'd give it, either.

File: runtime/ts/type.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (shans) [reply runtime/ts/type.ts]: Done.

**Pull request merged**